File: datasets/Python/Mistral/DataStructures/easy5.py
from collections import deque, defaultdict

def print_top_view():
    n = int(input())
    if n == 0:
        print()
        return

    nodes = list(map(int, input().split()))
    node_to_index = {val: idx for idx, val in enumerate(nodes)}
    root_val = nodes[0]

    # Build adjacency list for the tree
    adj = defaultdict(list)
    e = int(input())
    for _ in range(e):
        parent, child, lr = input().split()
        parent = int(parent)
        child = int(child)
        adj[parent].append((child, lr))

    # BFS setup: queue stores (node, HD)
    hd_map = {}
    queue = deque()
    queue.append((root_val, 0))

    while queue:
        node, hd = queue.popleft()
        if hd not in hd_map:
            hd_map[hd] = node

        for (child, lr) in adj.get(node, []):
            if lr == 'L':
                queue.append((child, hd - 1))
            else:
                queue.append((child, hd + 1))

    # Sort by HD and print
    for hd in sorted(hd_map.keys()):
        print(hd_map[hd], end=' ')
    print()

print_top_view()

File: datasets/C++/Mistral/DataStructures/easy5.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <algorithm>
using namespace std;

void printTopView() {
    int n;
    cin >> n;
    if (n == 0) {
        cout << endl;
        return;
    }

    vector<int> nodes(n);
    map<int, int> nodeToIndex;
    for (int i = 0; i < n; ++i) {
        cin >> nodes[i];
        nodeToIndex[nodes[i]] = i;
    }
    int rootVal = nodes[0];

    map<int, vector<pair<int, char>>> adj;
    int e;
    cin >> e;
    while (e--) {
        int parent, child;
        char lr;
        cin >> parent >> child >> lr;
        adj[parent].push_back({child, lr});
    }

    map<int, int> hdMap;  // HD -> node value
    queue<pair<int, int>> q;  // {node, HD}
    q.push({rootVal, 0});

    while (!q.empty()) {
        auto [node, hd] = q.front();
        q.pop();

        if (hdMap.find(hd) == hdMap.end()) {
            hdMap[hd] = node;
        }

        for (auto [child, lr] : adj[node]) {
            if (lr == 'L') {
                q.push({child, hd - 1});
            } else {
                q.push({child, hd + 1});
            }
        }
    }

    for (auto [hd, val] : hdMap) {
        cout << val << " ";
    }
    cout << endl;
}

int main() {
    printTopView();
    return 0;
}
